File: src/services/blockchain/event-monitor.ts
/**
 * Event Monitor
 * Monitors blockchain events in real-time and processes historical events
 */

import { config } from '../../config';
import { logger } from '../../utils/logger.util';
import { AppDataSource } from '../../database/data-source';
import { Transaction } from '../../database/entities/Transaction.entity';
import { TransactionType } from '../../utils/constants';
import { notificationService } from '../notification.service';
import { ProviderManager } from './provider.manager';
import { DepositProcessor } from './deposit-processor';
import Redis from 'ioredis';

// FIX #16: Redis client for tracking historical fetch state
const redis = new Redis({
  host: config.redis.host,
  port: config.redis.port,
  password: config.redis.password,
  db: config.redis.db,
});

export class EventMonitor {
  private isMonitoring = false;
  private reconnectAttempts = 0;
  private readonly MAX_RECONNECT_ATTEMPTS = 10;
  private readonly INITIAL_RECONNECT_DELAY_MS = 1000; // 1 second initial delay
  private readonly MAX_RECONNECT_DELAY_MS = 30000; // 30 seconds max delay
  private readonly HISTORICAL_BLOCKS_LOOKBACK = 2000; // ~100 minutes on BSC (3s per block)

  // FIX #16: Track if historical events were already fetched in this session
  private hasEverFetched = false;

  // FIX #16: Cooldown to prevent spam fetching on rapid reconnects
  private readonly FETCH_COOLDOWN_MS = 5 * 60 * 1000; // 5 minutes
  private readonly LAST_FETCH_KEY = 'blockchain:last_historical_fetch';

  constructor(
    private providerManager: ProviderManager,
    private depositProcessor: DepositProcessor
  ) {}

  /**
   * Start monitoring blockchain for deposits
   */
  public async startMonitoring(): Promise<void> {
    if (this.isMonitoring) {
      logger.warn('⚠️ Blockchain monitoring is already running');
      return;
    }

    try {
      await this.providerManager.initializeWebSocket();

      const usdtContractWs = this.providerManager.getUsdtContractWs();

      if (!usdtContractWs) {
        throw new Error('WebSocket USDT contract not initialized');
      }

      // Set WebSocket error handlers
      this.providerManager.setWebSocketErrorHandlers(
        () => this.handleWebSocketDisconnect(),
        () => this.handleWebSocketDisconnect()
      );

      // Listen for Transfer events to system wallet
      const filter = usdtContractWs.filters.Transfer(
        null,
        config.blockchain.systemWalletAddress
      );

      usdtContractWs.on(filter, async (from, to, value, event) => {
        try {
          this.providerManager.updateWsActivity(); // Update activity timestamp
          await this.depositProcessor.handleTransferEvent(from, to, value, event);
        } catch (error) {
          logger.error('❌ Error handling Transfer event:', error);
        }
      });

      // Start WebSocket health check
      this.providerManager.startWsHealthCheck(() => this.handleWebSocketDisconnect());

      // Start orphaned deposit cleanup job
      this.depositProcessor.startCleanupJob();

      // FIX #16: Only fetch historical on FIRST start (not reconnects)
      if (!this.hasEverFetched) {
        await this.fetchHistoricalEvents();
        this.hasEverFetched = true;
      } else {
        logger.info('⏭️ Skipping historical fetch (reconnect detected)');
      }

      this.isMonitoring = true;
      logger.info('✅ Blockchain monitoring started');
      logger.info(`📡 Listening for deposits to: ${config.blockchain.systemWalletAddress}`);
    } catch (error) {
      logger.error('❌ Failed to start blockchain monitoring:', error);
      throw error;
    }
  }

  /**
   * Stop monitoring blockchain
   */
  public async stopMonitoring(): Promise<void> {
    try {
      this.isMonitoring = false;

      // Stop WebSocket health check
      this.providerManager.stopWsHealthCheck();

      // Stop cleanup job
      this.depositProcessor.stopCleanupJob();

      // Destroy WebSocket provider
      await this.providerManager.destroyWebSocket();

      logger.info('✅ Blockchain monitoring stopped');
    } catch (error) {
      logger.error('❌ Error stopping blockchain monitoring:', error);
    }
  }

  /**
   * Handle WebSocket disconnect and reconnect
   */
  private async handleWebSocketDisconnect(): Promise<void> {
    if (!this.isMonitoring) {
      return; // Don't reconnect if monitoring is stopped
    }

    this.reconnectAttempts++;

    // Alert admins on critical reconnection issues
    await notificationService.alertWebSocketDisconnect(
      this.reconnectAttempts,
      this.MAX_RECONNECT_ATTEMPTS
    ).catch((err) => {
      logger.error('Failed to send WebSocket alert', { error: err });
    });

    if (this.reconnectAttempts > this.MAX_RECONNECT_ATTEMPTS) {
      logger.error('❌ Max reconnect attempts reached. Stopping monitoring.');
      this.stopMonitoring();
      return;
    }

    // Calculate exponential backoff delay: min(initialDelay * 2^attempts, maxDelay)
    const exponentialDelay = Math.min(
      this.INITIAL_RECONNECT_DELAY_MS * Math.pow(2, this.reconnectAttempts - 1),
      this.MAX_RECONNECT_DELAY_MS
    );

    logger.info(
      `🔄 Reconnecting WebSocket (attempt ${this.reconnectAttempts}/${this.MAX_RECONNECT_ATTEMPTS}) in ${exponentialDelay}ms...`
    );

    setTimeout(async () => {
      try {
        await this.providerManager.initializeWebSocket();
        await this.startMonitoring();
        this.reconnectAttempts = 0; // Reset on successful reconnect
        logger.info('✅ WebSocket reconnected successfully');
      } catch (error) {
        logger.error('❌ Failed to reconnect WebSocket:', error);
        this.handleWebSocketDisconnect();
      }
    }, exponentialDelay);
  }

  /**
   * Fetch and process historical Transfer events since last processed block
   * Called once on startup to catch any deposits made while bot was offline
   * FIX #16: Enhanced with Redis tracking and cooldown to prevent spam
   */
  private async fetchHistoricalEvents(): Promise<void> {
    try {
      // FIX #16: Check cooldown - prevent spam fetching on rapid reconnects
      const lastFetchTime = await redis.get(`${this.LAST_FETCH_KEY}:time`);
      if (lastFetchTime) {
        const timeSinceLastFetch = Date.now() - parseInt(lastFetchTime);
        if (timeSinceLastFetch < this.FETCH_COOLDOWN_MS) {
          const remainingMs = this.FETCH_COOLDOWN_MS - timeSinceLastFetch;
          logger.info('⏸️ Historical fetch on cooldown', {
            remainingSeconds: Math.round(remainingMs / 1000),
          });
          return;
        }
      }

      const transactionRepo = AppDataSource.getRepository(Transaction);
      const httpProvider = this.providerManager.getHttpProvider();
      const usdtContract = this.providerManager.getUsdtContract();
      const currentBlock = await httpProvider.getBlockNumber();

      // FIX #16: Get last fetched block from Redis (persists across restarts)
      // Fallback to database if Redis doesn't have it
      const lastFetchedBlockRedis = await redis.get(`${this.LAST_FETCH_KEY}:block`);
      let fromBlock: number;

      if (lastFetchedBlockRedis) {
        // Use Redis tracked block (most reliable)
        fromBlock = parseInt(lastFetchedBlockRedis) + 1;
        logger.debug('Using last fetched block from Redis', { fromBlock: fromBlock - 1 });
      } else {
        // Fallback to database
        const lastTransaction = await transactionRepo.findOne({
          where: { type: TransactionType.DEPOSIT },
          order: { block_number: 'DESC' },
          select: ['block_number'],
        });

        fromBlock = lastTransaction?.block_number
          ? Number(lastTransaction.block_number) + 1
          : currentBlock - this.HISTORICAL_BLOCKS_LOOKBACK;

        logger.debug('Using last block from database fallback', {
          fromBlock,
          hasTransaction: !!lastTransaction,
        });
      }

      const toBlock = currentBlock;

      if (fromBlock >= toBlock) {
        logger.info('✅ No historical blocks to process (already up to date)', {
          currentBlock,
          lastFetched: fromBlock - 1,
        });
        return;
      }

      logger.info(
        `🔍 Fetching historical Transfer events from block ${fromBlock} to ${toBlock} (${toBlock - fromBlock} blocks)...`
      );

      // Query historical Transfer events to system wallet
      const filter = usdtContract.filters.Transfer(
        null,
        config.blockchain.systemWalletAddress
      );

      const events = await usdtContract.queryFilter(filter, fromBlock, toBlock);

      if (events.length === 0) {
        logger.info('✅ No historical deposits found');

        // FIX #16: Still update tracking even if no events
        await redis.set(`${this.LAST_FETCH_KEY}:block`, String(toBlock));
        await redis.set(`${this.LAST_FETCH_KEY}:time`, String(Date.now()));

        return;
      }

      logger.info(`📥 Found ${events.length} historical Transfer events, processing...`);

      // Process events sequentially to maintain order
      let processed = 0;
      let skipped = 0;

      for (const event of events) {
        try {
          if (!event.args) continue;

          const [from, to, value] = event.args;

          // Check if already processed
          const existing = await transactionRepo.findOne({
            where: { tx_hash: event.transactionHash },
          });

          if (existing) {
            skipped++;
            continue;
          }

          await this.depositProcessor.handleTransferEvent(from, to, value, event);
          processed++;
        } catch (error) {
          logger.error(`❌ Error processing historical event ${event.transactionHash}:`, error);
        }
      }

      // FIX #16: Update last fetched block and time in Redis
      await redis.set(`${this.LAST_FETCH_KEY}:block`, String(toBlock));
      await redis.set(`${this.LAST_FETCH_KEY}:time`, String(Date.now()));

      logger.info(
        `✅ Historical events processed: ${processed} new, ${skipped} already processed, ${events.length} total`,
        {
          fromBlock,
          toBlock,
          blocksFetched: toBlock - fromBlock,
        }
      );
    } catch (error) {
      logger.error('❌ Error fetching historical events:', error);
      // Don't throw - allow monitoring to continue even if historical fetch fails
    }
  }

  /**
   * Get monitoring status
   */
  public isActive(): boolean {
    return this.isMonitoring;
  }
}

File: app/services/blockchain/event_monitor.py
"""
Event Monitor for USDT Transfer events.

Monitors blockchain for incoming USDT transfers to system wallet.
"""

import asyncio
from collections.abc import Callable
from decimal import Decimal

from loguru import logger
from web3 import AsyncWeb3

from .constants import USDT_ABI, USDT_DECIMALS


class EventMonitor:
    """
    Monitors USDT Transfer events on BSC blockchain.

    Uses polling mechanism to detect incoming transfers to monitored addresses.
    """

    def __init__(
        self,
        web3: AsyncWeb3,
        usdt_contract_address: str,
        poll_interval: int = 3,
    ) -> None:
        """
        Initialize event monitor.

        Args:
            web3: AsyncWeb3 instance
            usdt_contract_address: USDT contract address
            poll_interval: Polling interval in seconds
        """
        self.web3 = web3
        self.usdt_contract_address = web3.to_checksum_address(
            usdt_contract_address
        )
        self.poll_interval = poll_interval

        # Create contract instance
        self.usdt_contract = self.web3.eth.contract(
            address=self.usdt_contract_address,
            abi=USDT_ABI,
        )

        # Monitoring state
        self._monitoring = False
        self._last_processed_block = 0
        self._event_callback: Callable | None = None

        logger.info(
            f"EventMonitor initialized for USDT contract: "
            f"{self.usdt_contract_address}"
        )

    async def start_monitoring(
        self,
        watch_address: str,
        from_block: int | str = "latest",
        event_callback: Callable | None = None,
    ) -> None:
        """
        Start monitoring USDT transfers to address.

        Args:
            watch_address: Address to monitor for incoming transfers
            from_block: Starting block number or 'latest'
            event_callback: Async callback function for new transfers
        """
        self._monitoring = True
        self._event_callback = event_callback

        watch_address_checksum = self.web3.to_checksum_address(watch_address)

        # Get starting block
        if from_block == "latest":
            self._last_processed_block = await self.web3.eth.block_number
        else:
            self._last_processed_block = int(from_block)

        logger.info(
            f"Started monitoring USDT transfers to {watch_address_checksum} "
            f"from block {self._last_processed_block}"
        )

        # Start polling loop
        while self._monitoring:
            try:
                await self._poll_events(watch_address_checksum)
                await asyncio.sleep(self.poll_interval)

            except Exception as e:
                logger.error(f"Error in monitoring loop: {e}")
                await asyncio.sleep(self.poll_interval * 2)

    async def _poll_events(self, watch_address: str) -> None:
        """
        Poll for new Transfer events.

        Args:
            watch_address: Address to watch
        """
        try:
            # Get current block
            current_block = await self.web3.eth.block_number

            if current_block <= self._last_processed_block:
                return  # No new blocks

            # Get Transfer events filter
            # Filter transfers TO the watch address
            event_filter = (
                await self.usdt_contract.events.Transfer.create_filter(
                    from_block=self._last_processed_block + 1,
                    to_block=current_block,
                    argument_filters={"to": watch_address},
                )
            )

            # Get all events
            events = await event_filter.get_all_entries()

            # Process each event
            for event in events:
                await self._process_event(event)

            # Update last processed block
            self._last_processed_block = current_block

        except Exception as e:
            logger.error(f"Error polling events: {e}")

    async def _process_event(self, event: any) -> None:
        """
        Process a single Transfer event.

        Args:
            event: Web3 event object
        """
        try:
            # Extract event data
            from_address = event["args"]["from"]
            to_address = event["args"]["to"]
            value_wei = event["args"]["value"]

            # Convert to USDT (18 decimals)
            amount_usdt = Decimal(value_wei) / Decimal(10**USDT_DECIMALS)

            tx_hash = event["transactionHash"].hex()
            block_number = event["blockNumber"]

            logger.info(
                f"Detected USDT transfer: {amount_usdt} USDT\n"
                f"  From: {from_address}\n"
                f"  To: {to_address}\n"
                f"  TX: {tx_hash}\n"
                f"  Block: {block_number}"
            )

            # Call event callback if provided
            if self._event_callback:
                await self._event_callback(
                    {
                        "from_address": from_address,
                        "to_address": to_address,
                        "amount": float(amount_usdt),
                        "tx_hash": tx_hash,
                        "block_number": block_number,
                    }
                )

        except Exception as e:
            logger.error(f"Error processing event: {e}")

    async def stop_monitoring(self) -> None:
        """Stop event monitoring."""
        self._monitoring = False
        logger.info("Event monitoring stopped")

    @property
    def is_monitoring(self) -> bool:
        """Check if monitoring is active."""
        return self._monitoring
